### ISSUE TITLE
http3: reject unsupported headers

### ISSUE THREAD

**Issue by bt90** [quic-go/quic-go#4647]
I think we're still allowing a few headers which are violating [RFC9113](https://datatracker.ietf.org/doc/html/rfc9113#name-connection-specific-header-) and [RFC9114](https://datatracker.ietf.org/doc/html/rfc9114#name-http-fields)

* `Connection`
* `TE` with any other value as `trailers`
* `Keep-Alive`
* `Proxy-Connection`
* `Upgrade`

According to the spec they should result in a stream error of type [H3_MESSAGE_ERROR](https://datatracker.ietf.org/doc/html/rfc9114#H3_MESSAGE_ERROR)

**Comment by marten-seemann** [comment]
RFC 9113 doesn't apply to HTTP/3, does it?

**Comment by bt90** [comment]
[RFC9114 Section 4.2](https://datatracker.ietf.org/doc/html/rfc9114#name-http-fields) doesn't list them explicitly, but refers to them as `connection-specific header fields` and links to [RFC911 Section 7.6.1](https://datatracker.ietf.org/doc/html/rfc9110#name-connection).

**Comment by bt90** [comment]
It may seem obvious, but we should also throw an error if we are running as a client and these headers are passed to the request. We shouldn't rely on a third-party HTTP/3 server implementation to do the right thing. This could avoid some bugs caused by triggering spec-violating behaviour.

**Comment by marten-seemann** [comment]
> It may seem obvious, but we should also throw an error if we are running as a client and these headers are passed to the request.

That only works for the client though. There's no way to do this kind of check in a server request handler, due to the lack of an error return value on the `http.ResponseWriter.WriteHeader`. All we can do is ignore (and not serialize) invalid header fields. I'm not sure if I like this asymmetry.

**Comment by bt90** [comment]
I guess the QUIC layer is completely out of scope at this point?

**Comment by marten-seemann** [comment]
What do you mean? We can do stream resets, but that doesn't help communicate to the server that it's doing something wrong.

**Comment by bt90** [comment]
Terminating the QUIC stream with `H3_MESSAGE_ERROR` would be the correct thing to do according to the spec.

**Comment by marten-seemann** [comment]
Sure, that's what we should do.

I thought you want to return an error to HTTP users when they _write_ (and not only when they receive) the invalid values, but maybe I misunderstood your proposal.

**Comment by bt90** [comment]
Now I get it. This client/server receive/respond mix tends to get a bit confusing from time to time.

So we're able to properly reject them as we receive them as part of an incoming request, but things get ugly once somebody would try to craft a malformed request?

**Comment by marten-seemann** [comment]
Turns out we already have logic to skip these headers: https://github.com/quic-go/quic-go/blob/1a25053e5eb59d53eca46bbc726e5935b7ce2f3c/http3/request_writer.go#L147-L154

This is in code that we copied from the standard library, which is probably why I didn't remember that it's there ;)



> So we're able to properly reject them as we receive them as part of an incoming request, but things get ugly once somebody would try to craft a malformed request?

Yes of course. That's why this issue is included in the next milestone. The sanitization logic makes it difficult to test this end-to-end, but at least we can have some unit tests.
